File: Centre_College/CSC_341_FA2014/Projects/CPlusPlus/radixSort.cpp
/**
 * CSC 341 Assignment 2
 * radixSort.cpp
 * Purpose: This program was our first in c++. This program uses a 
 * linked list to sort numbers in a text document ("data.txt") using 
 * a radix sort algorithm. 
 * 
 * @author Daniel Graham
 * @version 1.0 10/24/2014
 * */

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

using namespace std;

/**
 * Implements a Node object with integers as values and a link to the 
 * next Node in series. Will be used for the Queue implementation later.
 * */

class Node{
	int value;
	Node * link;
  public:
	Node();
	Node (int);
	void setValue(int);
	void setLink(Node *);
	Node* getLink();
	void print();
	int getValue();
};

/**
 * Constructs a generic Node. Does not have a link, and the default
 * value is -1.
 * */

Node :: Node(){
	value = -1;
}

/** 
 * Constructs a Node to a specified value. Does not have a link.
 * 
 * @param startNum The initial value of the value variable for a Node.
 * */

Node :: Node(int startNum){
	value = startNum;
}

/**
 * Changes the value of the Node object from its initial value to num.
 * 
 * @param num The new value of the value variable.
 * */

void Node :: setValue(int num){
	value = num;
}

/**
 * Returns the linked Node of the current Node.
 * 
 * @return link The Node that the current Node has stored as link.
 * */

Node* Node :: getLink(){
	return link;
}

/**
 * Returns the value stored in the Node.
 * 
 * @return value The value stored in the Node.
 * */

int Node :: getValue(){
	return value;
}

/**
 * Prints the value stored in the Node to the screen.
 * */

void Node :: print(){
	cout << getValue() << "\n";
}

/**
 * Sets the link of the current Node to a new Node pointer. 
 * 
 * @param nodeToLink The pointer to the Node which will be linked to
 * 					 the current Node.
 * */
 
void Node :: setLink(Node * nodeToLink){
	link = nodeToLink;
}

/**
 * Implements an integer queue with the preceeding Node objects. Allows 
 * basic enqueue and dequeue options, but is limited to integers.
 * */

class Queue{
		int length;
		Node * queueHead, * queueTail;
	public:
		Queue();
		void enqueue(int);
		int dequeue();
		int peek(Node *);
		void printQueue();
		void radixSort();
		void readFromFile();
	};
	
/**
*   Constructs an empty Queue with a length of 0.
* */
	
Queue :: Queue(){
	length = 0;
}

/**
 * 	Enqueues an integer onto the Queue using a Node. Also sets the 
 * 	pointers to the tail (and head in an initially empty list) to the 
 * 	correct positions. The method also increments the length variable 
 * 	and sets links correctly.
 * 
 * @param numToEnqueue The number to enqueue to the Queue.
 * */
	
void Queue :: enqueue(int numToEnqueue){
	Node * nodeToEnqueue = new Node(numToEnqueue);
	
	if(length == 0){
		queueHead = nodeToEnqueue;
		queueTail = nodeToEnqueue;
	}
	else{
		(*queueTail).setLink(nodeToEnqueue);
		queueTail = nodeToEnqueue;
	}
	length ++;
}

/**
 * 	Dequeues the Node at the queueHead position, unless Queue is empty.
 * 	Returns the value at the head of the Queue, and deletes the pointer
 * 	to that Node.
 * 
 * @return returnValue Either NULL if the Queue is empty, or the integer
 * 					   at the head of the Queue.
 * */

int Queue :: dequeue(){
	Node * newHead;
	if (length == 0){
		cout << "Queue EMPTY!\n";
		return NULL;
	}
	else{
		int returnValue = (*queueHead).getValue();
		newHead = (*queueHead).getLink();
		delete queueHead;
		queueHead = newHead;
		length --;
		return returnValue;
	}
}

/**
 * 	Looks at the value of a Node in the Queue without dequeueing it.
 * 	Used in the printQueue() method.
 * 
 * 	@param nodeToPeek The pointer of the Node to get the value of.
 * 	@return Returns the value of the Node at the pointer in param.
 * */
 
int Queue :: peek(Node * nodeToPeek){
	return (*nodeToPeek).getValue();
}

/**
 * 	Prints the Queue in order without changing the Queue.
 * */
 
void Queue :: printQueue(){
	int staticLength = length;
	for(int i = 0; i < staticLength; i ++){
		int numToDequeue = dequeue();
		cout << numToDequeue << "\n";
		enqueue(numToDequeue);
	}
}

/**
 * 	Sorts the Queue using the Radix algorithm. Uses a tempSortQueue to 
 * 	store sorts from each tens place sort. Only sorts numbers up to 
 * 	4 digits long. 
 * */

void Queue :: radixSort(){
	Queue* tempSortQueue = new Queue;
	int staticLength = length;
	for(int i = 1; i < 1001; i = i * 10){
		for(int j = 0; j < 10; j++){
			staticLength = length;
			for (int k = 0; k < staticLength; k++){
				if(length > 0){
					int numToTest = dequeue();
					int numToTestModified = numToTest / i;
					if((numToTestModified % 10) == j){
						(*tempSortQueue).enqueue(numToTest);
					}
					else{
					enqueue(numToTest);
					}
				}
			}
		}
		while((*tempSortQueue).length > 0){
			int oneSortNum = (*tempSortQueue).dequeue();
			enqueue(oneSortNum);
		}
	}
	delete tempSortQueue;
}

/**
 * 	Reads the file "data.txt" and enqueues the items in that file in
 * 	the order they are present in that file. Stops input if reaches a
 * 	negative number or the end of the file.
 * */
 	
void Queue :: readFromFile(){
	//Code inspired by: http://www.cplusplus.com/doc/tutorial/files/
		string line;
		ifstream inputFile;
		inputFile.open("data.txt");
		int intFromFile;
		while(getline(inputFile, line)){
			stringstream convert(line);
			if( !(convert >> intFromFile)){
				intFromFile = 0;
				cout << "Could not convert File to integers";
			}
			else if(intFromFile >= 0){
				enqueue(intFromFile);
			}
			else{
				break;
			}
		}
	}
	
/**
 * This method tests the radixSort, Queue, and Node	objects.
 * **/
		

int main(){
	
	
	/*The commented out code was used for testing.
	Node node1(1),node2(2),node3(3), returnNode;
	node1.print();
	node2.print();
	node3.print();
	returnNode.print();
	node1.setValue(4);
	node1.print();
	node1.linkNode(&node2);
	node2.linkNode(&node3);
	returnNode.setValue((*(node2.getNextNode())).getValue());
	returnNode.print();
	
	/*
	 * OUTPUT SHOULD BE:
	 * 1
	 * 2
	 * 3
	 * -1
	 * 4
	 * 3
	 * 
	cout << "\n";
	Queue testQueue, testQueue1, testQueue2;
	testQueue.enqueue(4);
	testQueue.enqueue(3);
	testQueue.enqueue(2);
	testQueue.enqueue(1);

	testQueue.printQueue();
	testQueue.dequeue();
	testQueue.dequeue();
	cout << "\n";
	testQueue.printQueue();
	/*
	 * OUTPUT SHOULD BE:
	 * 4
	 * 3
	 * 2
	 * 1
	 * 
	 * 2
	 * 1
	 * 
	cout << "\n";
	testQueue.enqueue(3);
	testQueue.enqueue(4);
	testQueue.radixSort();
	testQueue.printQueue();
	
	/*Output should be:
	 * 1
	 * 2
	 * 3
	 * 4
	 * 
	cout << "\n";
	testQueue.enqueue(324);
	testQueue.enqueue(6346);
	testQueue.enqueue(1000);
	testQueue.enqueue(5);
	testQueue.radixSort();
	testQueue.printQueue();
	
	/* Output should be:
	 * 1
	 * 2
	 * 3
	 * 4
	 * 5
	 * 324
	 * 1000
	 * 6346
	 * 
	
	
	/*TEST FOR READ FROM FILE
	
	cout << "\n";
	cout << "\n";
	cout << "\n";
	
	testQueue2.readFromFile();
	testQueue2.radixSort();
	testQueue2.printQueue();
	
	*/
	Queue sortedQueue;
	sortedQueue.readFromFile();
	sortedQueue.radixSort();
	sortedQueue.printQueue();
	
	return 0;
}

File: Centre_College/CSC_341_FA2014/Projects/Fortran/radixSort.f
c A program to implement Radix sort in Fortran 77
c This was done for CSC 341.
c Sorts the numbers in file "data.txt"
c By Daniel Graham
c version 1.1 10/17/2014


c Takes an array of size 101 and prepares it to act as a queue of
c integers by setting every address of the array to the value -1
c The head of the queue is not always array(1), but changes as items
c are enqueued and dequeued. The queueHead holds the index of the start
c of the queue as it changes. The last item always comes before a -1
c but theres is also a queueEnd that holds the index of the last item.
c
c Params: queue The 101 size array to use as a queue.

      subroutine initializeQueue(queueHead, queueTail, queue, queueSize)
      
      integer queue(*), queueHead, queueTail, queueSize
      queueHead = 1
      queueTail = 1
      queueSize = 0
      do 60 i = 1, 101
        queue(i) = -1
 60   continue
      return
      end

c Reads a .txt line by line for positive integers. The integers must be
c no more than 4 digits and there cannot be more than 100 positive
c integers. Increments a count of integers and inserts the integers to
c an array.
c
c Params: numElements Number of elements in the list.
c         fName File path of .txt which contains integers to add.
c         array The array that will contain integers.

      subroutine inputArray(numElements, fName, array)

      Character fName*(*)
      integer numElements, array(*)
      
      open(44, File = fName, Status = 'OLD')
      read(44, *) line
      i = 1

c The following loop iterates over the file and if the number it reaches
c is non-negative, it increments numElements and adds the number to the 
c array otherwise, it stops reading the file.

 10   if(line.GE.0) then
          numElements = numElements + 1
          array(numElements) = line
          read(44, *) line
          goto 10
      endif
      close(44)
      return
      end

c Enqueues an integer to the queue array. If the queue already has 100
c items, the subroutine sends a message. Changes the value of queueEnd 
c to match the updated end of the queue. 
c
c Params: queueTail Index of the next open space.
c         queue Array that has been initialized to a queue.
c         item Integer to add to the queue.
c         queueSize Current size of the queue

      subroutine enqueue(queueTail, queue, item, queueSize)

      integer queueTail, queue(*), item, queueSize
      
      if(queue(queueTail).NE.-1) then
        write(*,*) "Queue Full!"
      else
        queue(queueTail) = item
        queueSize = queueSize + 1
        
        if(queueTail.GE.101) then
            queueTail = 1
        else
            queueTail = queueTail + 1
        endif
        
      endif

      return
      end

c Returns the integer at the head of the queue. If the 
c queue is empty prints an error to the screen and returns -1. 
c Decrements queueSize.
c
c Params: queueHead Location of the item at the head of the queue.
c         queue The queue array to dequeue from.
c         queueSize The current size of the queue array.  
      
      integer function dequeue(queueHead, queue, queueSize)

      integer queueHead, queue(*), queueSize
      
      if(queue(queueHead).EQ.-1) then
        write(*,*) "No Items On Queue"
        dequeue = -1
      else
        dequeue = queue(queueHead)
        queue(queueHead) = -1
        queueSize = queueSize - 1
        
        if(queueHead.GE.101) then
            queueHead = 1
        else
            queueHead = queueHead + 1
        endif
        
      endif
      return
      end
      
c Uses radix sort to sort the integers in the queue by a specific place 
c (10, 100, etc.). Sorts by creating a temporary queue to sort into, 
c then returning items to the original queue in the sorted order.
c
c Params: place The place to sort by.
c         queue The queue array to sort.
c         queueHead The index of the head of the queue array.
c         queueTail Index of the next available location in queue
c         queueSize Current size of queue.

      subroutine oneSort(place, queue, queueHead, queueTail, 
     + queueSize)
     
      integer queueSort(101), queue(*)
      integer place, queueHead, queueSortHead
      integer queueTail, queueSortTail, queueSize
      integer queueSortSize, dequeueItem, modValue, dequeue, toMod
      
      call initializeQueue(queueSortHead, queueSortTail, queueSort,
     + queueSortSize)

      do 70 i = 0, 9
        do 80 j = 1, (queueSize)
            dequeueItem = dequeue(queueHead, queue, queueSize)
            toMod = dequeueItem / place
            modValue = mod(toMod,10)
            if(modValue.EQ.i) then
                call enqueue(queueSortTail, queueSort, dequeueItem, 
     + queueSortSize)
            else
                call enqueue(queueTail, queue, dequeueItem,
     + queueSize)
            endif
c This if statement determines where in the sort queue items will go

 80     continue
 70    continue
       do 90 i = 1, queueSortSize
        call enqueue(queueTail, queue, 
     + dequeue(queueSortHead, queueSort, queueSortSize),queueSize)
 90    continue
       return
       end

c Prints the queue to console in order seperated by commas.
c
c Params: queueHead The index of the head of the queue array.
c         queue The queue array to sort.
c         queueSize Current size of queue.

      subroutine printQueue(queueHead, queue, queueSize)

      integer queueHead, queueSize, queue(*), printPoint
      integer queueToPrint(101), queueToPrintHead
      integer queueToPrintTail, queueToPrintSize
      
      call initializeQueue(queueToPrintHead, queueToPrintTail, 
     + queueToPrint, queueToPrintSize)

      printPoint = queueHead
      do 110 i = 1, 100
        if(printPoint.GT.101) then
            printPoint = 1
        endif
        call enqueue(queueToPrintTail, queueToPrint,  
     + queue(printPoint), queueToPrintSize)
      printPoint = printPoint + 1
 110  continue
      write(*,*) queueToPrint
      return
      end
      
c Sorts integers in "data.txt" file using radix sort algorithm and 
c queue implementation. Cannot sort more than 100 integers and the 
c integers cannot be more than 4 digits. Prints the sorted output to 
c the screen. Ignore tailing -1s.      
      
      program radixSort
      integer queue(101), queueHead, queueTail, queueSize
      Character fileName*(100)
      
      fileName = "data.txt"
      call initializeQueue(queueHead, queueTail, queue, queueSize)
      call inputArray(queueSize, fileName, queue)
      queueTail = queueSize + 1
      
      call oneSort(1, queue, queueHead, queueTail,
     +  queueSize)
      call oneSort(10, queue, queueHead, queueTail,
     +  queueSize)
      call oneSort(100, queue, queueHead, queueTail,
     +  queueSize)
      call oneSort(1000, queue, queueHead, queueTail,
     +  queueSize)
c I could have done the above as a loop, but is clearer as is.

      call printQueue(queueHead, queue, queueSize)
      
      stop
      end
      
      
      
